### PR TITLE
MGMT-12894: Use the hub release image when determining the ironic agent image

### PR DESCRIPTION
**Pull request by carbonin**
Previously, when using the converged flow, the service attempted to extract the ironic agent image using the spoke release. This was not ideal for a few reasons:

1. The agent version should match that of the hub ironic service
    - While it was likely the agent from the spoke version would be compatible with the hub service it's objectively better for them to actually match.
2. The default image was used for late-binding cases
    - This meant that a rather old and seemingly random image was used when the service didn't have a cluster from which to pull a release image
3. It's difficult to communicate to a user what to mirror in disconnected cases

With this change the agent will always match the hub service, the hub agent will be used regardless of late-binding or not, and the user just needs to know to mirror the hub release into their spoke environment when disconnected.

The only case when we should use the default image now is if the hub release image doesn't support the CPU architecture of the spoke infraEnv, but with multi-arch images coming soon and differing hub/spoke architectures exceedingly rare already this shouldn't be a problem.

## List all the issues related to this PR

Resolves https://issues.redhat.com/browse/MGMT-12894

- [ ] New Feature 
- [x] Enhancement 
- [ ] Bug fix
- [ ] Tests
- [ ] Documentation
- [ ] CI/CD 

## What environments does this code impact?

- [ ] Automation (CI, tools, etc)
- [ ] Cloud
- [x] Operator Managed Deployments
- [ ] None

## How was this code tested?

1. Deployed this change in a dev-scripts environment.
2. Created a bmh referencing an infraenv (late binding)
3. Downloaded discovery ignition for the infraenv
4. Validated the ironic agent image in the ignition matched the one from the hub release image

- [ ] assisted-test-infra environment
- [x] dev-scripts environment
- [ ] Reviewer's test appreciated
- [x] Waiting for CI to do a full test run
- [x] Manual (Elaborate on how it was tested)
- [ ] No tests needed

## Checklist

- [x] Title and description added to both, commit and PR.
- [x] Relevant issues have been associated (see [CONTRIBUTING] guide)
- [ ] This change does not require a documentation update (docstring, `docs`, README, etc)
- [x] Does this change include unit-tests (note that code changes require unit-tests)

## Reviewers Checklist

- Are the title and description (in both PR and commit) meaningful and clear?
- Is there a bug required (and linked) for this change?
- Should this PR be backported?

### CODE REVIEW

**Comment by carbonin** [comment]
Marking this as WIP because I want to run it through manual tests, but I also want to see CI run on it.

**Comment by carbonin** [comment]
Ah I guess the converged job will be broken until https://github.com/openshift/assisted-service/pull/4779 merges.

**Comment by carbonin** [comment]
/hold

Putting this on hold as I remembered that we wanted to keep the spoke release image behavior for compatibility.
As it is now this would require users who are currently installing in disconnected environments to mirror new images which we probably don't want.

cc @eranco74 @filanov what do you think?

**Comment by filanov** [comment]
> /hold
> 
> Putting this on hold as I remembered that we wanted to keep the spoke release image behavior for compatibility. As it is now this would require users who are currently installing in disconnected environments to mirror new images which we probably don't want.
> 
> cc @eranco74 @filanov what do you think?

we can keep using the spoke image if we don't have access to the hub version. if it's always accessible i see no reason in using the spoke (besides the architecture issue) 
Muli arch images will probably be delayed to 4.14

**Comment by carbonin** [comment]
>if it's always accessible i see no reason in using the spoke

I was thinking about upgrade cases (and I thought we talked about this, but I could be thinking about something else).

Say there is a user installing 4.11 with ZTP disconnected. Right now they just need to mirror the spoke release image, but when they upgrade they'll also need to mirror the hub. Do we care about making them make this change? I'm fine with putting it in a release note, but if we want to keep this compatibility then we'd need to keep the spoke logic.

File: internal/controller/controllers/preprovisioningimage_controller.go [hunk 1]
```diff
@@ -419,26 +418,22 @@ func (r *PreprovisioningImageReconciler) AddIronicAgentToInfraEnv(ctx context.Co
 	return ctrl.Result{}, err
 }
 
-func (r *PreprovisioningImageReconciler) getIronicAgentImageByRelease(ctx context.Context, log logrus.FieldLogger, infraEnv common.InfraEnv) (string, error) {
-	cluster, err := r.Installer.GetClusterInternal(ctx, installer.V2GetClusterParams{ClusterID: infraEnv.ClusterID})
-	if err != nil {
+func (r *PreprovisioningImageReconciler) getIronicAgentImageByRelease(ctx context.Context, log logrus.FieldLogger, infraEnv *common.InfraEnv) (string, error) {
+	cv := &configv1.ClusterVersion{}
+	if err := r.Get(ctx, types.NamespacedName{Name: "version"}, cv); err != nil {
```
Comment (carbonin) [inline internal/controller/controllers/preprovisioningimage_controller.go:423]: Maybe it's better to do this once at startup and provide the image to the controller when we create it :thinking:

Comment (carbonin) [reply internal/controller/controllers/preprovisioningimage_controller.go:423]: It's not like the hub is going to change images after our pod starts

File: internal/controller/controllers/preprovisioningimage_controller.go [hunk 2]
```diff
@@ -420,26 +421,32 @@ func (r *PreprovisioningImageReconciler) AddIronicAgentToInfraEnv(ctx context.Co
...
+		architectures, err = r.OcRelease.GetReleaseArchitecture(log, cv.Status.Desired.Image, r.ReleaseImageMirror, infraEnv.PullSecret)
+		if err != nil {
+			return "", err
+		}
+		image, err = r.OcRelease.GetIronicAgentImage(log, cv.Status.Desired.Image, r.ReleaseImageMirror, infraEnv.PullSecret)
+		if err != nil {
+			return "", err
+		}
+		r.hubIronicAgentImage = image
+		r.hubReleaseArchitectures = architectures
```
Comment (eranco74) [inline internal/controller/controllers/preprovisioningimage_controller.go:442]: Add logging

**Review by eranco74 — COMMENTED** [review]
looks good, 
Though the converged flow and disconnected jobs are failing

**Comment by carbonin** [comment]
> fix lint

Works for me locally. Will retest.

**Comment by carbonin** [comment]
Some of these test failures will be addressed in https://github.com/openshift/assisted-service/pull/4822

File: docs/hive-integration/README.md [hunk 3]
```diff
@@ -215,7 +215,7 @@ The environment var for the ironicAgent image to be used on arm64 CPU architectu
 `IRONIC_AGENT_IMAGE_ARM`
 
 **NOTE**
-Ensure the correct images are mirrored if installing in a disconnected environment
+Ensure the correct images are mirrored if installing in a disconnected environment.
```
[the remark this answers is not in the document]
Comment (carbonin) [reply docs/hive-integration/README.md:218]: Added some more detail. Let me know what you think.

**Comment by carbonin** [comment]
/unhold

This is ready, but I wasn't able to sort out why the converged test is failing. We should probably wait on merging until that passes.

**Comment by eranco74** [comment]
The converged periodic job is green
https://prow.ci.openshift.org/job-history/gs/origin-ci-test/logs/periodic-ci-openshift-assisted-service-release-ocm-2.6-e2e-ai-operator-ztp-converged-periodic

**Comment by eranco74** [comment]
Looks like the agent never started (the node didn't register)
We don't seem to gather logs from hosts that didn't register.

**Comment by carbonin** [comment]
Ran this through manual testing again in my dev-scripts environment and the agent discovers fine.

I grabbed the ironic agent image spec it's trying to use and I can't seem to pull it locally:
```
16:12:43:~$ podman pull quay.io/openshift-release-dev/ocp-v4.0-art-dev@sha256:920ffbcfe14f5b2122d1fe7c079fa1d45df63c0e6b50ce9c953dcc2d482bcc82
Trying to pull quay.io/openshift-release-dev/ocp-v4.0-art-dev@sha256:920ffbcfe14f5b2122d1fe7c079fa1d45df63c0e6b50ce9c953dcc2d482bcc82...
Error: initializing source docker://quay.io/openshift-release-dev/ocp-v4.0-art-dev@sha256:920ffbcfe14f5b2122d1fe7c079fa1d45df63c0e6b50ce9c953dcc2d482bcc82: reading manifest sha256:920ffbcfe14f5b2122d1fe7c079fa1d45df63c0e6b50ce9c953dcc2d482bcc82 in quay.io/openshift-release-dev/ocp-v4.0-art-dev: unauthorized: access to the requested resource is not authorized
```

It looks like the cluster image set we're trying to use for the spoke is `registry.build05.ci.openshift.org/ci-op-i61vsc6m/release@sha256:578d960e87d2250e1dd35fb88762cd8355fb88dafdd4a8353b01bcaf786b3964` which is internal to the build farm.

Maybe the agent is having trouble pulling the hub image?
Are we using a different pull secret for the spoke than we did to install the hub?
Is the hub being installed through some kind of mirror?

**Comment by carbonin** [comment]
The dev-scripts setup logs are also showing an internal release image used for the install:

```
2023-01-03 14:18:47 ++(common.sh:153): source(): OPENSHIFT_RELEASE_IMAGE=registry.build05.ci.openshift.org/ci-op-i61vsc6m/release@sha256:578d960e87d2250e1dd35fb88762cd8355fb88dafdd4a8353b01bcaf786b3964
```

And the service logs show that this was correctly used to fetch the ironic agent:
```
time="2023-01-03T15:16:24Z" level=info msg="Fetching image from OCP release (oc adm release info --image-for=ironic-agent --insecure=false registry.build05.ci.openshift.org/ci-op-i61vsc6m/release@sha256:578d960e87d2250e1dd35fb88762cd8355fb88dafdd4a8353b01bcaf786b3964)" func="github.com/openshift/assisted-service/internal/oc.(*release).getImageFromRelease" file="/assisted-service/internal/oc/release.go:291" go-id=707 preprovisioning_image=ostest-extraworker-0 preprovisioning_image_namespace=assisted-spoke-cluster request_id=bf2f4799-c9df-4da0-9247-f640cbb2a848
time="2023-01-03T15:16:25Z" level=info msg="Caching hub ironic agent image quay.io/openshift-release-dev/ocp-v4.0-art-dev@sha256:920ffbcfe14f5b2122d1fe7c079fa1d45df63c0e6b50ce9c953dcc2d482bcc82 for architectures [x86_64]" func="github.com/openshift/assisted-service/internal/controller/controllers.(*PreprovisioningImageReconciler).getIronicAgentImageByRelease" file="/assisted-service/internal/controller/controllers/preprovisioningimage_controller.go:443" go-id=707 preprovisioning_image=ostest-extraworker-0 preprovisioning_image_namespace=assisted-spoke-cluster request_id=bf2f4799-c9df-4da0-9247-f640cbb2a848
```

I also checked that this matches the cluster version for good measure.:
```
  desired:
    image: registry.build05.ci.openshift.org/ci-op-i61vsc6m/release@sha256:578d960e87d2250e1dd35fb88762cd8355fb88dafdd4a8353b01bcaf786b3964
    version: 4.12.0-0.nightly-2023-01-02-175114
```

Is this not valid in the CI system for some reason?
@osherdp any ideas?

**Comment by osherdp** [comment]
@carbonin what's the symptom you're seeing here? host of spoke cluster not coming back from boot?

**Comment by carbonin** [comment]
The agent never registers.

My guess is that there's some failure to pull the ironic agent image (which runs before our agent), but nothing is collected from those spoke hosts.

**Comment by osherdp** [comment]
> The agent never registers.
> 
> My guess is that there's some failure to pull the ironic agent image (which runs before our agent), but nothing is collected from those spoke hosts.

can you (at the time being) rerun and debug the job live? should be easy enough to investigate from there (although a bit rigorous) 
we have some task to do sos-reporting for those VMs, but I think it will take a while before we'll have it ready

**Comment by carbonin** [comment]
> can you (at the time being) rerun and debug the job live?
Yup, this was my plan for today, but just haven't gotten around to it yet.

Can you confirm that we use the same pull secret for the hub and spoke and both should have connectivity to wherever we're getting the images from? I'm just wondering if there's something obvious that I'm missing like a mirror registry/proxy/auth issue.

**Comment by osherdp** [comment]
> > can you (at the time being) rerun and debug the job live?
> 
> Yup, this was my plan for today, but just haven't gotten around to it yet.
> 
> Can you confirm that we use the same pull secret for the hub and spoke and both should have connectivity to wherever we're getting the images from? I'm just wondering if there's something obvious that I'm missing like a mirror registry/proxy/auth issue.

hub and spoke should have the same connectivity (as they are both implemented as VMs in the same libvirt networks) and we're not suppose to have two values of a pull-secret for them

**Comment by carbonin** [comment]
:shrug: it passed this time

I didn't do anything when I was debugging it, I don't know what was different.

**Comment by eranco74** [comment]
@osherdp @adriengentil can we gather sosreport from all hosts, so we can tell why a host didn't registered?

**Comment by osherdp** [comment]
> 

as I said. it's WIP

**Pull request merged**